Add resume policy instructions for Katib experiments (#2324)
kubeflow/website: content/en/docs/components/hyperparameter-tuning/resume-experiment.md

                    
+++

> This page describes in detail how to modify running experiment
> and restart succeeded experiment. Follow this guide to know more
> about changing the experiment execution process and use various

Reviewer: - "a running experiment" or "running experiment"
- "page" -> "guide" (everything can be a page?)
- "succeeded"? -> "completed" (success in ML experiments can be subjective, but completion (it's over) is more objective, I think)
- Talk to the reader - "you will learn more..."

```suggestion
This guide describes how to modify running experiments
and restart completed experiments. You will learn
about changing the experiment execution process and use various
```

resume policies for the Katib experiment.

> For details of how to configure and run your experiment, see the guide to

Reviewer: - Grammar
- a11y

```suggestion
For details on how to configure and run your experiment, check the guide on
```

Reviewer: Also, check the links please (e.g. `/docs/components/hyperparameter-tuning/experiment/` since some may be moving to `.../kativ/`.. I think?)

## Modify running experiment

> While the experiment is running you are able to change trial count parameters.
> For example, if you want to decrease the maximum number of
> hyperparameter sets that are trained parallel.

Reviewer: This may appear like an incomplete sentence because of the "if" statement. Let's try the following:

```suggestion
While the experiment is running you are able to change trial count parameters.
For example, you can decrease the maximum number of
hyperparameter sets that are trained in parallel.
```

Note: hyperparam sets can be trained "in parallel" not "parallel", I think.

kubectl edit experiment <experiment-name> -n <experiment-namespace>
```

Make appropriate changes and save it. Controller automatically processes
new parameters and makes necessary changes.

- If you want to increase or decrease parallel trial execution,
  modify `parallelTrialCount`. Controller accordingly creates or
  deletes trials in line with `parallelTrialCount`.

- If you want to increase or decrease maximum trial count,
  modify `maxTrialCount`. `maxTrialCount` should be greater than current
  count of `Succeeded` trials. You can remove this parameter, if
  the experiment should run endless with `parallelTrialCount` parallel trials
  until it reaches `Goal` or `maxFailedTrialCount`.

- If you want to increase or decrease maximum failed trial count,
  modify `maxFailedTrialCount`. You can remove this parameter,
  if the experiment should not reach `Failed` status.

## Resume succeeded experiment

To control various resume policies, you can specify `.spec.resumePolicy`
for the experiment.

> See the [`ResumePolicy` type](https://github.com/kubeflow/katib/blob/master/pkg/apis/controller/experiments/v1beta1/experiment_types.go#L54).

Reviewer: ```suggestion
(Refer to the [`ResumePolicy` type](https://github.com/kubeflow/katib/blob/master/pkg/apis/controller/experiments/v1beta1/experiment_types.go#L54).)
```

Katib experiment is restartable only if it is in **`Succeeded`** status because
`maxTrialCount` has been reached. To check current experiment status run:
`kubectl get experiment <experiment-name> -n <experiment-namespace>`.

To restart experiment, you are able to change only **`parallelTrialCount`**,
**`maxTrialCount`** and **`maxFailedTrialCount`**
as described [above](#modify-experiment)

### Resume policy: Never

Use this policy if your experiment should not be resumed at any time.
After the experiment has succeeded,
the suggestion's [deployment](https://kubernetes.io/docs/concepts/workloads/controllers/deployment/)
and [service](https://kubernetes.io/docs/concepts/services-networking/service/)
are deleted and you can't restart the experiment.

> Read more about Katib concepts in the

Reviewer: For a11y:

```suggestion
Learn more about Katib concepts in the
```

[overview guide](/docs/components/hyperparameter-tuning/overview/#katib-concepts).

> See the [never resume policy example](https://github.com/kubeflow/katib/blob/master/examples/v1beta1/resume-experiment/never-resume.yaml#L20).

Reviewer: For a11y:

```suggestion
Check the [`never-resume.yaml`](https://github.com/kubeflow/katib/blob/master/examples/v1beta1/resume-experiment/never-resume.yaml#L20)
example for more details.
```

WDYT?

You can omit `.spec.resumePolicy` parameter for that functionality.

### Resume policy: FromVolume

Use this policy if you intend to restart the experiment.
In that case, [volume](https://kubernetes.io/docs/concepts/storage/volumes/)
is attached to the suggestion's deployment.

Katib controller creates
[persistent volume claim](https://kubernetes.io/docs/concepts/storage/persistent-volumes/#persistentvolumeclaims)
(PVC) and
[persistent volume](https://kubernetes.io/docs/concepts/storage/persistent-volumes/#persistent-volumes)
(PV) in addition to the suggestion's deployment and service.

**Note:** If you specify
[storage class name](https://kubernetes.io/docs/concepts/storage/storage-classes/)
in Katib configuration settings for the PVC differently from `katib-suggestion`,
Katib controller creates only PVC. Follow the
[Katib configuration guide](/docs/components/hyperparameter-tuning/katib-config/#suggestion-volume-settings)
to set up the suggestion's volume settings.

- PVC is deployed with the name: `<suggestion-name>-<suggestion-algorithm>`
  in the suggestion namespace.

- PV is deployed with the name:
  `<suggestion-name>-<suggestion-algorithm>-<suggestion-namespace>`

> After the experiment has succeeded, the suggestion's deployment and

Reviewer: ```suggestion
After the experiment is successful, the suggestion's deployment and
```

or maybe "has finished"? "Successful" can be subjective, IMHO

service are deleted. Suggestion data can be retained in the volume.
When you restart the experiment, the suggestion's deployment and service
are created and suggestion statistics can be recovered from the volume.

When you delete the experiment, the suggestion's deployment, service,
PVC and PV are deleted automatically.

> See the
> [from volume policy example](https://github.com/kubeflow/katib/blob/master/examples/v1beta1/resume-experiment/from-volume-resume.yaml#L18).

Reviewer: for a11y:

```suggestion
Check the
[`from-volume-resume.yaml`](https://github.com/kubeflow/katib/blob/master/examples/v1beta1/resume-experiment/from-volume-resume.yaml#L18)
example to learn more.
```

WDYT? It's more precise, since it's not a "tutorial" example, just a "hands-on" YAML file.

## Next steps

- Learn how to [configure and run your Katib experiments](/docs/components/katib/experiment/).

> - Read about [Katib Configuration (Katib config)](/docs/components/katib/katib-config/).

Reviewer: for a11y:

```suggestion
- Check the [Katib configuration (Katib config)](/docs/components/katib/katib-config/) page.
```
